### ISSUE TITLE
Auto adding parentheses for function calls

### ISSUE THREAD

**Issue by hd2015** [microsoft/vscode-cpptools#882]
During auto-completion for methods, brackets do not get inserted automatically. Most of the time they are required and it's easier to delete them than insert them manually. Calling methods is common than passing them as arguments.

Eg: Tab or Return after "size_t n = name.len" should autocomplete to "size_t n = name.length()" rather than "size_t n = name.length" with the cursor between/after (according to whether there is function argument)  the brackets.

**Comment by bobbrow** [comment]
I think this idea has merit, but it's not something that we can implement in the extension.  You would have to ask VS Code to implement it.

We do not have control of the cursor position and if we supplied the completion result in your example, the cursor would end up after the closing paren in every case.  This would be ok for methods with no arguments, but not for methods requiring arguments.  Even if we could put the cursor in the right place, that would mess up VS Code's notifications for parameter hints since they expect the user to type the opening paren before they will initiate a request for parameter hints.

**Comment by Gurpartap** [comment]
There's a way to do this with certain language extensions (typescript, go). Details here: https://github.com/Microsoft/vscode-go/issues/276#issuecomment-205439716

**Comment by IgorOne** [comment]
@bobbrow  I'm sorry but this is BS, Several other language extensions (Go, TS, JS, C#) added this and it is a must for C/C++. And actually, the VS Code team does not plan to do this, I quote "we defer it to the language service" from here: https://github.com/Microsoft/vscode/issues/1021#issuecomment-171893794 .

So there is no one other than the cpp team that can give us this functionality.

**Comment by sean-mcmanus** [comment]
I'm seeing typescript invoking completion when ( is typed? Is that the desired result or something else?

I think we need to add ( as a "commit character" for methods (i.e. https://github.com/OmniSharp/omnisharp-vscode/issues/1487 ). It's part of the CompletionItem type in the latest language server protocol, but the one we use appears to be old and is missing it (https://microsoft.github.io/language-server-protocol/specification ).

**Comment by IgorOne** [comment]
I actually tried to look through the code but I didn't even manage to find the part where the autocompletion happens. So I hope someone can either implement this or point to the code which is responsible for completion.

**Comment by sean-mcmanus** [comment]
@IgorOne It's implemented in our closed-source Microsoft.VSCode.CPP.Extension.* binary. I agree that we should fix this, but I'm not sure of an ETA yet (seems almost more like a bug than a feature request).

**Comment by sean-mcmanus** [comment]
Our latest [0.21.0-insiders3](https://github.com/Microsoft/vscode-cpptools/releases) triggers completion when `(` is used. Is that good enough or is adding the () on Enter/Tab/Click, still desired?

Setting `C_Cpp.updateChannel` to `"Insiders"` is another way to get the update.

**Comment by jacekkk** [comment]
Coming from JetBrains world, I'd also find this feature extremely useful. I'd much rather see it implemented in a way that @hd2015 describes, so as soon as you press enter to accept a suggestion, a pair of brackets is added for me (provided it's a function call, not a variable) and cursor is placed either inside of after them (which is what happens in JetBrains IDEs). What @sean-mcmanus  talks about is already present in VSCode (at least in my version), but it's not exactly what we're looking for as all it does is adding a closing bracket as soon as you type opening one, which has nothing to do with IntelliSense and code completion.

**Comment by sean-mcmanus** [comment]
@JacasEdi The change we added was that typing "(" commits the completion. The only difference is typing ( to finish the completion versus typing tab/enter. I'll keep this issue open to track the tab/enter method.

UPDATE: We have a fix, but it won't be in 0.22.0-insiders, because we intend to add a setting that controls how the parenthesis completion.

**Comment by foreverlms** [comment]
@sean-mcmanus Is this problem solved? The autocompletion of brackets is really in-need.

**Comment by sean-mcmanus** [comment]
@foreverlms Yeah, we just need to add some setting like C_Cpp.methodCompletion: "AddNoParens", "AddLeftParen", "AddBothParens". The work to do that has been post-poned till later.

**Comment by foreverlms** [comment]
@sean-mcmanus Thanks! And I also want to ask that how can I automatically turn `.` after pointers into `->`? I haven't found a specific answer. Appreciate for your reply.

**Comment by sean-mcmanus** [comment]
@foreverlms You found the issue tracking that at https://github.com/Microsoft/vscode-cpptools/issues/862.

**Comment by Raymoun17** [comment]
is it fixed yet ? coming from PyCharm this feature is a must for me

**Comment by sean-mcmanus** [comment]
@Raymoun17 Not fixed yet -- needs settings added.

**Comment by ranqiyu** [comment]
yes, this feature is very important. from visual studio 2017 and use c++ ' user.

**Comment by SashiRin** [comment]
This feature is very important and I also want this feature, too.

**Comment by risalfajar** [comment]
Still waiting for this feature to be implemented. It is this feature that keeps me using CLion instead of VSCode.

**Comment by 0xDedSec42** [comment]
It would be really awesome to get this feature.

**Comment by hhc97** [comment]
Agreed, this feature would be excellent to have.

**Comment by akrsnr** [comment]
It can be solved by ticking `javascript.suggest.completeFunctionCalls` property up. Here is the gif.

https://gfycat.com/cleveractualhornbill

**Comment by Raymoun17** [comment]
finally

**Comment by hhc97** [comment]
@akrsnr Yes, but that only works for Javascript and not C.

**Comment by dodalovic** [comment]
@hhc97  it does work in js, ts, but it doesn't work in `js` inside `html`:

```html
<script>
  document.getElementById
</script>
```

Here `getElementById` didn't get the parenthesis added :(

**Comment by hhc97** [comment]
@dodalovic What I mean is that it does not work when coding in the C programming language.

**Comment by Mohammad-Yasser** [comment]
Almost 3 years old issue and still no fix. :(

**Comment by nixmeer** [comment]
This missing feature is the reason I prefer atom with the autocomplete-clang package. It is much more comfortable... but it's still atom -.-

**Comment by xsjqqq123** [comment]
Wait this feature and " . to -> "

**Comment by GrethTol** [comment]
Any news from devs? Is it contains in any roadmap? Because this is must have feature :F

**Comment by Er2620** [comment]
Has this feature been added yet?

**Comment by michelleangela** [comment]
@Er2620 
This feature has not been added. It's currently planned to be worked on after version 1.0.0 of the extension.

**Comment by Er2620** [comment]
> @Er2620
> This feature has not been added. It's currently planned to be worked on after version 1.0.0 of the extension.

Ok, thank you for letting me know.

**Comment by superRaptor911** [comment]
status of this feature?

**Comment by michelleangela** [comment]
@superRaptor911 the status of this feature request has not changed. It's under the milestone that's post version 1.0.0 of the extension.

**Comment by gvcallen** [comment]
Would really like this feature and surprised its taking so long.

**Comment by samuelyhsu** [comment]
This feature is really important, please

**Comment by Geometror** [comment]
I would like this essential feature too!

**Comment by Lawstorant** [comment]
Ridiculous. Such an essential and easy to implement feature but it takes ages because of ***

*Commented moderated. We can understand frustration, but please keep it respectful here.  [Microsoft Code of Conduct](https://opensource.microsoft.com/codeofconduct/)*

**Comment by kterhorst** [comment]
Please MS! Just implement this essential feature that makes so many coders happy.

**Comment by sean-mcmanus** [comment]
Yeah, I had an implementaton in cpptools from Feb 2019, but we needed some setting to control the behavior, such as C_Cpp.methodCompletion: "AddNoParens", "AddLeftParen", "AddBothParens"...no one submitted a PR to add those settings.

I'm sort of busy working on Windows ARM64 support...trying not to get distracted by random feature requests.

@bobbrow Was there a reason this was moved to Backlog? It seems like it'd be a quick fix.

**Comment by bobbrow** [comment]
We haven't lost sight of this.  It has sufficient upvotes and will be included in a future milestone.  But we can only work on so many things at a time.  We will update the milestone when it has been scheduled for a release.  Thank you for your patience.

**Comment by xfarmer** [comment]
For python, `"python.autoComplete.addBrackets": true,` worked for me

**Comment by Jianchaoliang** [comment]
unsubscribe


| |
菅朝梁
|
|
18205601716@163.com
|
签名由网易邮箱大师定制
On 11/3/2020 10:55，Farmer Li<notifications@github.com> wrote：

For python, "python.autoComplete.addBrackets": true, worked for me

**Comment by abhijoshi2k** [comment]
Every text editor has this feature and not this one! I'm surprised and disappointed. This is the only reason of not using vscode.

**Comment by sean-mcmanus** [comment]
@bobbrow This has 47 votes now -- seems pretty high to me.

**Comment by sean-mcmanus** [comment]
Auto adding parentheses is available with https://github.com/microsoft/vscode-cpptools/releases/tag/1.3.0-insiders . It's enabled by default (C_Cpp.autocompleteAddParentheses). The behavior of when to add the closing `)` also responds to the editor.autoClosingBrackets setting (default is to always close the `)`). Let us know if you find any issues related to this.

**Comment by xiray** [comment]
> Auto adding parentheses is available with https://github.com/microsoft/vscode-cpptools/releases/tag/1.3.0-insiders . It's enabled by default (C_Cpp.autocompleteAddParentheses). The behavior of when to add the closing `)` also responds to the editor.autoClosingBrackets setting (default is to always close the `)`). Let us know if you find any issues related to this.

Thanks for the enhancement, it helps a lot!
In using the feature, I found 2 annoying unexpected auto completions, please take a look:
```
class Foo {
public:
  Foo(int n) : n_(n) {}
  virtual ~Foo() {}
};

class Bar : public Foo {
public:
  void func() {
    Foo* a = nullptr; // Autocomplete Foo to "Foo()" is unexpected in typing 'Foo'
  }

private:
  vector<Foo*> x_; // Autocomplete Foo to "Foo()" is unexpected in typing 'Foo'
};
```

**Comment by sean-mcmanus** [comment]
@xiray Thanks for reporting that. I've moved it to https://github.com/microsoft/vscode-cpptools/issues/7209. It's fixed for our pending 1.3.0-insiders2 (along with another issue involving template methods), but we're still trying to decide what do in cases when the method name is supposed to be used without `()` and/or the IO stream manipulator case (https://github.com/microsoft/vscode-cpptools/issues/7199).

**Comment by sean-mcmanus** [comment]
With https://github.com/microsoft/vscode-cpptools/releases/tag/1.3.0-insiders2, the C_Cpp.autocompleteAddParentheses setting defaults to false (so you'd need to set it to true to enable it). We also fixed a couple bugs with it.

**Comment by michelleangela** [comment]
This is now available in the 1.3.0 release version https://github.com/microsoft/vscode-cpptools/releases/tag/1.3.0
